zapisy/apps/offer/plan/utils.py

```diff
@@ -0,0 +1,508 @@
+# As an argument, it takes a single dictionary entry from get_votes function.
+def propose(vote: ProposalVoteSummary):
+    current_year = SystemState.get_current_state().year
+    proposal = Proposal.objects.get(id=vote.proposal)
+    avg = SingleVote.objects.filter(state__year=current_year, value__gt=0).values('proposal').annotate(
+        total=Sum('value')).aggregate(Avg('total'))
+    previous_avg = 0
+    years = 0
+    percentage = 0.8
+
+    for year, values in vote.voting.items():
+        if not year == current_year:
+            years += 1
+            previous_avg = values['total']
+    if proposal.course_type.obligatory:
+        return True
+    if proposal.recommended_for_first_year:
+        return True
+    if vote.voting[current_year]['total'] >= avg['total__avg']:
+        return True
+    if years > 0 and vote.voting[current_year]['total'] >= percentage * (previous_avg / years):
+        return True
+    return False
+
+
+def get_subjects_data(subjects: List[Tuple[str, str, int]], years: int):
+    # Function prepares data to create a necessary csv to assign classes. Given a course, it will look for previous instances
+    # of the course (taking the newest one), and copy the group assigments from it. If such instance does not exist, it will
+    # assign only the owner of the course.
+    # Arguments:
+    #   subjects: list of tuples (course name, course semester, course proposal id). Each course represented
+    #             by course name will be present in return value.
+    #   years: tells the function how many years back should it look for previous instances of
+    #          the course to copy data from.
+    # Return value: list of list of pairs. Each entry is a single group. Each pair is of (name, value) format:
+    # | name      | desc                                   |
+    # | course    | course name                            |
+    # | semester  | either z or l                          |
+    # | teacher   | group teacher                          |
+    # | code      | teacher code                           |
+    # | type      | group type (lecture, lab etc.)         |
+    # | hours     | hours allocated for this type of class |
+
+    course_data = {}
+    states = get_last_years(years)
+
+    for subject in subjects:
+        course_data[subject[0]] = {'id': subject[2], 'semester': subject[1],
+                                   'instance': CourseInstance.objects.filter(
+            reduce(lambda x, y: x | y, [Q(semester__year=year) for year in states]), offer=subject[2],
+            semester__type=subject[1]).order_by('-semester__year')}
+
+    groups = []
+    # Prepares data necessary to create csv.
+    for course, data in course_data.items():
+        proposal_info = Proposal.objects.get(id=data['id'])
+        if data['instance']:
+            previous_groups = Group.objects.filter(course=data['instance'][0])
```

Comment:
Dzięki temu `first()` już nie trzeba robić takich paskudnych rzeczy.
```suggestion
            previous_groups = Group.objects.filter(course=data['instance'])
```